qutip/solver/propagator.py

```diff
@@ -82,29 +90,93 @@ def propagator(
         tlist = [0, t]
         list_output = False
     else:
-        tlist = t
+        tlist = list(t)
         list_output = True
 
     if not isinstance(H, (Qobj, QobjEvo)):
         H = QobjEvo(H, args=args, **kwargs)
 
     if isinstance(c_ops, list):
         c_ops = [QobjEvo(op, args=args, **kwargs) for op in c_ops]
+    elif c_ops is not None and not isinstance(c_ops, QobjEvo):
+        c_ops = [QobjEvo(c_ops, args=args, **kwargs)]
+
+    c_ops_list: list[QobjEvo] = []
+    if isinstance(c_ops, QobjEvo):
+        c_ops_list = [c_ops]
+    elif isinstance(c_ops, list):
+        c_ops_list = c_ops
+
+    constant_c_ops = all(op.isconstant for op in c_ops_list)
+    H_isconstant = isinstance(H, QobjEvo) and H.isconstant
+
+    if piecewise_t is not None:
+        if not constant_c_ops:
+            raise ValueError(
+                "piecewise_t is provided for H, but collapse operators are not constant"
+            )
+        if H_isconstant:
+            warnings.warn(
+                "piecewise_t provided but the Hamiltonian is time-independent",
+                UserWarning,
+            )
+        piecewise_times = [
+            tt for tt in piecewise_t if tlist[0] < tt <= tlist[-1]
+        ]
+        times = sorted(set(tlist + piecewise_times))
+        target = set(tlist)
+        H_step = H(tlist[0], args)
+        if c_ops_list:
+            c_ops_q = [op(tlist[0], args) for op in c_ops_list]
+            gen = liouvillian(H_step, c_ops_q)
+        else:
+            gen = H_step if H_step.issuper else -1j * H_step
```

Comment:
I think it's more correct as
```suggestion
            gen = H_step if H_step.issuper else liouvillian(H_step, [])
```